src/Components/WebAssembly/WebAssembly/src/Hosting/WebAssemblyCultureProvider.cs

```diff
@@ -63,45 +65,31 @@ public void ThrowIfCultureChangeIsUnsupported()
 
     public virtual async ValueTask LoadCurrentCultureResourcesAsync()
     {
-        var culturesToLoad = GetCultures(CultureInfo.CurrentCulture);
-
-        if (culturesToLoad.Count == 0)
+        if (!OperatingSystem.IsBrowser())
         {
-            return;
+            throw new PlatformNotSupportedException("This method is only supported in the browser.");
         }
 
-        // Now that we know the cultures we care about, let WebAssemblyResourceLoader (in JavaScript) load these
-        // assemblies. We effectively want to resovle a Task<byte[][]> but there is no way to express this
-        // using interop. We'll instead do this in two parts:
-        // getSatelliteAssemblies resolves when all satellite assemblies to be loaded in .NET are fetched and available in memory.
-#pragma warning disable CS0618 // Type or member is obsolete
-        var count = (int)await _invoker.InvokeUnmarshalled<string[], object?, object?, Task<object>>(
-            GetSatelliteAssemblies,
-            culturesToLoad.ToArray(),
-            null,
-            null);
-
-        if (count == 0)
+        var culturesToLoad = GetCultures(CultureInfo.CurrentCulture);
+
+        if (culturesToLoad.Length == 0)
         {
             return;
         }
 
-        // readSatelliteAssemblies resolves the assembly bytes
-        var assemblies = _invoker.InvokeUnmarshalled<object?, object?, object?, object[]>(
-            ReadSatelliteAssemblies,
-            null,
-            null,
-            null);
-#pragma warning restore CS0618 // Type or member is obsolete
+        await WebAssemblyCultureProviderInterop.LoadSatelliteAssemblies(culturesToLoad, LoadSatelliteAssembly);
```

Comment:
Also to help me understand the JS interop a bit more, could you comment on the lifetime of the delegate that we're creating from `LoadSatelliteAssembly` here? I know the JS side receives a `(wrapper: { dll: Uint8Array }) => void` but how long is the .NET-side delegate pinned for? It's not obvious where this gets disposed/released. Is it released synchronously after the source-generated .NET code runs?

Reply:
No, not synchronously, because the interop code can't know if the JS code would keep reference to the JS function. 
We use `FinalizationRegistry` to do the cleanup together with `GCHandle`.